### PR TITLE
Make checkStackSize symbol weak.

### PR DESCRIPTION
**Pull request by zlobober**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Non-significant (changelog entry is not required)

There is a sanity check implemented by `checkStackSize()` call. It works fine when interpreters are instantiated by ClickHouse code in properly prepared threads, but there are cases when ClickHouse runs as a library inside another application. If application is using user-space lightweight threads with manually allocated stacks, current implementation is not reasonable, as it has no way to properly check the remaining stack size without knowing the details of how stacks are allocated.

Fortunately, `checkStackSize()` is a global routine, so it is pretty easy to make it customizable by declaring it [weak](https://en.wikipedia.org/wiki/Weak_symbol), so that it can be replaced by linker with a more suitable implementation.

This PR adds macro for marking symbol weak and marks `checkStackSize()` as weak.

### CODE REVIEW

**Comment by zlobober** [comment]
Comment about macro naming: Currently, ClickHouse does not prefix its macro with any prefix, so it may form a clash with macro names from other projects (as happened with LOG_INFO before), so I'd prefer to keep macro name `CH_WEAK` rather than simply `WEAK`. I believe that prefixing macro names with something unique is a good practice for code that may be used as a library.

File: dbms/src/Core/Defines.h [hunk 1]
```diff
@@ -162,3 +162,12 @@
 /// A macro for suppressing warnings about unused variables or function results.
 /// Useful for structured bindings which have no standard way to declare this.
 #define UNUSED(...) (void)(__VA_ARGS__)
+
+/// A macro for declaring global symbols weak. Useful when certain global function
+/// may be substituted by different implementation by another application using
+/// ClickHouse as a library (i.e. checkStackSize()).
+#if defined(__GNUC__)
```
Comment (alexey-milovidov) [inline dbms/src/Core/Defines.h]: We use attribute weak without any macro and it works fine both for `clang` and `gcc`; on Linux, FreeBSD, MacOS; on x86_64, aarch64, ppc64...

Just proceed without macro or replace all other usages of `__attribute__((__weak__))` in code.

**Comment by alexey-milovidov** [comment]
Ok.

File: dbms/src/Core/Defines.h (hunk 1 above)
Comment (zlobober) [reply dbms/src/Core/Defines.h]:
```
~/clickhouse/source/dbms $ fgrep -r "((__weak__))" .
~/clickhouse/source/dbms $ 
```

I did not find any occurrences of __attribute__((__weak__) in DBMS source code, can you point me to any of them.

Regarding the macro: current version is based on the Y_WEAK macro from Arcadia, I believe that __GNUC__ macro works for both g++ and clang, and it was introduced for compatibility with MSVC. Is ClickHouse buildable under MSVC? If not, I'll remove macro and proceed with raw __attribute__((__weak__)).

**Comment by alexey-milovidov** [comment]
``
./libs/libmemcpy/memcpy.h:void * __attribute__((__weak__)) memcpy(void * __restrict destination, const void * __restrict source, size_t size)
./libs/libcommon/src/DateLUTImpl.cpp:__attribute__((__weak__)) extern bool inside_main;
./libs/libcommon/src/LineReader.cpp:    char * readline(const char *) __attribute__((__weak__));
```

**Comment by alexey-milovidov** [comment]
> Is ClickHouse buildable under MSVC?

No.

**Comment by alexey-milovidov** [comment]
Ok. I also moved the explanation inside code.

**Pull request merged**